### ISSUE TITLE
Lack of Support for Retrieving the Currently Selected Audio Device and Defaulting to External Speaker on Certain Cases

### ISSUE THREAD

**Issue by huantd0111** [videosdk-live/videosdk-rtc-flutter-sdk-example#47]
Currently, I am working with the Video SDK integrated into a Flutter project for real-time communication. The implementation is generally successful; however, I have encountered an issue related to audio devices.

### Problem:
The Video SDK does not provide a method to retrieve the currently selected audio device, such as the earpiece, speakerphone, or Bluetooth headset, which is essential for users to know which device is currently in use.

**Defaulting to External Speaker:**
Furthermore, I noticed that in some specific scenarios, the app defaults to the external speaker even when other audio devices are available and selected. This issue particularly affects users who prefer using earpieces or Bluetooth headsets for privacy or convenience reasons.

### Expected Behavior:

The Video SDK should expose a method or API that enables developers to retrieve information about the currently selected audio device.
The app should respect the user's selected audio device and not automatically default to the external speaker unless explicitly chosen.
Steps to Reproduce:

Launch the app on various devices.
Connect different audio devices such as earpieces, Bluetooth headsets, or external speakers.
Observe the audio output device that the app defaults to in each scenario.
Devices and Environments Affected:
This issue has been observed on the following devices and environments:

Xiaomi Mi 9T (Android 10)
Samsung Galaxy S10 (Android 11)
Google Pixel 4 (Android 12)
iOS devices running iOS 14 and 15
Additional Information:
I have investigated the official documentation and forums, but there are no available solutions or workarounds to address this problem. As this issue can significantly impact user experience, I hope the development team can prioritize adding support for retrieving the currently selected audio device and improving the handling of default audio output.

### Suggested Solution:

Introduce a new method in the Video SDK that enables developers to obtain information about the currently selected audio device (e.g., "getCurrentAudioDevice()").
Review the audio handling logic to ensure the app respects the user's selected audio device and does not automatically default to the external speaker without explicit user input.
Your prompt attention to this matter would be greatly appreciated. Thank you for your dedication to enhancing the Video SDK and delivering a seamless real-time communication experience.

If there are any additional details or testing I can provide to assist with resolving this issue, please let me know.

Thank you!

**Comment by arjun-kava** [comment]
Hey, @huantd0111 Thanks for brief. We will look into it and get back to you soon.

**Comment by ishabodiwala** [comment]
Hey @huantd0111, Thank you for raising the issue. 
1. **Provide a method to retrieve the currently selected audio device** - we will definitely take this in backlog and update you for the same.
2. **Defaulting to External Speaker**  - we tested on several devices, but we could not reproduce this issue. Although we will test on other devices as well. meanwhile can you update the SDK version to `1.1.4` and check ?

**Comment by huantd0111** [comment]
Hello @ishabodiwala 
Thank you for your support. Regarding the "Defaulting to External Speaker" issue, I suspect that the problem of audio being played through the external speaker might be related to the microphone issue in version 1.1.3. In some cases, when I mute the microphone, the audio might still be transmitted to the other side. After unmuting the mic, the audio on the other device becomes very low.

To investigate this further, I will update the SDK to version 1.1.4 and conduct thorough testing. Once I've tested it on the latest SDK version, I'll provide you with additional information.

Thank you for considering this in the backlog, and I'll keep you updated after testing on the new SDK version.

Best regards,

**Comment by ishabodiwala** [comment]
@huantd0111, yes please test on the latest SDK version and let us know the results. We will be awaiting your update.

**Comment by ishabodiwala** [comment]
Hey @huantd0111, have you tested with latest SDK version ?

**Comment by huantd0111** [comment]
@ishabodiwala I apologize for keeping you waiting. However, due to some reasons, I could only access the real device and test with the latest SDK version recently. I have conducted the testing, and it seems that the issue did not appear in the latest SDK version.

**Comment by ishabodiwala** [comment]
@huantd0111, No problem! I understand it took some time, but the good news is that the latest SDK version fixed the issue. If you faced this issue again or need help with anything else, just let us know! You can join our discord community also.
https://discord.com/invite/Qfm8j4YAUJ

**Comment by huantd0111** [comment]
Retrieve currently selected audio device is not available on latest version v1.1.5

**Comment by nimisane** [comment]
I am facing the same issue

on some devices getAudioOutputDevices() method is not listing all the available audio output devices and even after switching to speakerphone from bluetooth connected earphones, still audio can is coming from earphone in between for some spit seconds

VideoSkd Version used
videosdk: ^1.1.5

Code:-

onTapDown: (details) async {
        List<MediaDeviceInfo>
            outptuDevice = _room!
                .getAudioOutputDevices();
        final screenSize =
            MediaQuery.of(context).size;

        await showMenu(
          context: context,
          color: AppColors.whiteFD,
          shape: RoundedRectangleBorder(
              borderRadius:
                  BorderRadius.circular(
                      12)),
          position:
              RelativeRect.fromLTRB(
                  25,
                  MediaQuery.of(context)
                          .padding
                          .top +
                      20,
                  screenSize.width,
                  0),
          items: outptuDevice.map((e) {
            return PopupMenuItem(
                value: e,
                child: Text(e.label));
          }).toList(),
          elevation: 8.0,
        ).then((value) {
          if (value != null) {
            _room!.switchAudioDevice(
                value);
          }
        });
      },

**Comment by ishabodiwala** [comment]
Hey @nimisane, on which device are you testing ?

**Comment by ZoofSoftwareSolutions** [comment]
@ishabodiwala  I have tested on iPhone 11, Samsung m14 and realme 5 pro till now

**Comment by neilhosey** [comment]
Is there anyway to set the external microphone to not be SPEAKER?

**Comment by ishabodiwala** [comment]
Hey @neilhosey, We're sorry for any inconvenience, but if your device isn't listed, it means our SDK doesn't currently recognise that specific device. Unfortunately, you won't be able to use that device at the moment. This particular issue demands thorough research. Right now, our team is actively working on implementing various features as per our roadmap. We'll be able to dedicate time to addressing this problem during our upcoming sprint.

**Comment by arjun-kava** [comment]
Resolved here: https://docs.videosdk.live/react/guide/video-and-audio-calling-api-sdk/setup-call/precall
